### PR TITLE
Pr thermometer algorithm issue 1159

### PR DESCRIPTION
**Pull request by xavierraffin**
Fix issue https://github.com/CanalTP/navitia/issues/1159

I did it accordingly with TeXitoi proposal.

### CODE REVIEW

File: source/time_tables/thermometer.cpp [hunk 1]
```diff
@@ -130,12 +132,72 @@ void Thermometer::generate_thermometer(const type::Route* route) {
...
+typedef std::vector< Vertex > Container;
+
+// Generate an topologicaly exact thermometer
+// return false if topological sort fail (because of a cycle)
+// return true if succeed
+bool Thermometer::generate_topological_thermometer(const std::vector<vector_idx> &stop_point_lists) {
+
+    Graph stop_point_graph;
+    std::map<type::idx_t, Vertex> vertex_map;
+    std::map<type::idx_t, Vertex>::iterator it;
```
Comment (TeXitoi) [inline source/time_tables/thermometer.cpp]: You can forget this line, and declare `auto it = ...` when you do your search.

File: source/time_tables/thermometer.cpp [hunk 2]
```diff
@@ -130,12 +132,72 @@ void Thermometer::generate_thermometer(const type::Route* route) {
...
+    for(auto v : stop_point_lists) {
```
Comment (TeXitoi) [inline source/time_tables/thermometer.cpp]: for consistency, we put a space between `for` and `(`. Same for `if`, `switch`, `while` and such keywords.

Comment (TeXitoi) [inline source/time_tables/thermometer.cpp]: here, you don't want to copy your vector, thus, it's most of the time a good idea do write your for loop using const reference:

``` c++
for (const auto& v: stop_point_list) {
```

File: source/time_tables/thermometer.cpp [hunk 3]
```diff
@@ -130,12 +132,72 @@ void Thermometer::generate_thermometer(const type::Route* route) {
...
+    Container c;
+    try {
+        boost::topological_sort(stop_point_graph, std::back_inserter(c));
+    }catch(const boost::not_a_dag &){
+        // Graph contains cycles : we need to try custom algorithm
+        return false;
+    }
+
+    // Build thermometer with sort result
+    for ( Container::reverse_iterator ii=c.rbegin(); ii!=c.rend(); ++ii) {
```
Comment (TeXitoi) [inline source/time_tables/thermometer.cpp]: you can use auto here:

``` c++
for (auto it = c.rbegin(); it != c.rend(); ++it) {
```

(and space around `=`, `==` and other comparison operators.

File: source/time_tables/thermometer.cpp [hunk 4]
```diff
@@ -130,14 +132,72 @@ void Thermometer::generate_thermometer(const type::Route* route) {
...
+                current_vx = boost::add_vertex(VertexProperties(sp), stop_point_graph);
+                vertex_map[sp] = current_vx;
+            } else {
+	        current_vx = it->second;
+	    }
+	    if (!first) {
+	        boost::add_edge(previous_vx, current_vx, stop_point_graph);
+	    }
+	first = false;
+	previous_vx = current_vx;
```
Comment (TeXitoi) [inline source/time_tables/thermometer.cpp]: strange indentation

File: source/time_tables/thermometer.cpp [hunk 5]
```diff
@@ -130,14 +132,72 @@ void Thermometer::generate_thermometer(const type::Route* route) {
...
-    nb_branches = 0;
-    std::vector<vector_idx> req;
-    if(stop_point_lists.size() > 1) {
-        for(auto v : stop_point_lists) {
-            if(req.empty())
+    if (stop_point_lists.size() > 1) {
+        // Try a topological_sort first
+        //If succeed, return, else use custom algorithm
+        if (generate_topological_thermometer(stop_point_lists))
+            return;
```
Comment (TeXitoi) [inline source/time_tables/thermometer.cpp]: We avoid if statement without brackets, but you can put everything on one line if you prefere, ie

``` c++
if (generate_topological_thermometer(stop_point_lists)) { return; }
```

or

``` c++
if (generate_topological_thermometer(stop_point_lists)) {
    return;
}
```

File: source/time_tables/thermometer.cpp [hunk 6]
```diff
@@ -130,14 +132,72 @@ void Thermometer::generate_thermometer(const type::Route* route) {
...
+        uint32_t max_sp = get_max_sp(stop_point_lists);
+        nb_branches = 0;
+        std::vector<vector_idx> req;
+        for (auto v : stop_point_lists) {
+            if (req.empty())
```
Comment (TeXitoi) [inline source/time_tables/thermometer.cpp]: brackets please

File: source/time_tables/thermometer.cpp [hunk 7]
```diff
@@ -175,9 +235,9 @@ std::vector<uint32_t> Thermometer::stop_times_order(const type::VehicleJourney&
 std::vector<uint32_t> Thermometer::stop_times_order_helper(const vector_idx &stop_point_list) const {
     std::vector<uint32_t> result;
     auto it = thermometer.begin();
-    for(type::idx_t spidx : stop_point_list) {
+    for (type::idx_t spidx : stop_point_list) {
         it = std::find(it, thermometer.end(),  spidx);
-        if(it==thermometer.end())
+        if (it == thermometer.end())
```
Comment (TeXitoi) [inline source/time_tables/thermometer.cpp]: brackets please

File: source/time_tables/thermometer.cpp [hunk 8]
```diff
@@ -191,14 +251,14 @@ std::vector<uint32_t> Thermometer::stop_times_order_helper(const vector_idx &sto
 vector_idx Thermometer::generate_possibilities(const std::vector<vector_idx> &journey_patterns, std::vector<vector_size> &pre_computed_lb) {
 
     //C'est qu'il n'y a pas de possibilités possibles
-    if(journey_patterns[0].empty() && journey_patterns[1].empty())
+    if (journey_patterns[0].empty() && journey_patterns[1].empty())
```
Comment (TeXitoi) [inline source/time_tables/thermometer.cpp]: brackets

**Comment by TeXitoi** [comment]
Great job! :+1:

File: source/time_tables/thermometer.cpp [hunk 9]
```diff
@@ -130,14 +132,72 @@ void Thermometer::generate_thermometer(const type::Route* route) {
...
 
+// Define types for next function 'generate_topological_thermometer'
+struct VertexProperties
+{
+    type::idx_t idx;
+    VertexProperties() : idx(0) {}
+    VertexProperties(type::idx_t const& i): idx(i) {}
+};
+typedef boost::adjacency_list< boost::vecS, boost::vecS, boost::directedS, boost::property<boost::vertex_bundle_t, VertexProperties> > Graph;
+typedef boost::graph_traits<Graph>::vertex_descriptor Vertex;
```
Comment (xlqian) [inline source/time_tables/thermometer.cpp]: prefer doing `using newType = oldType`

File: source/time_tables/thermometer.cpp [hunk 10]
```diff
@@ -130,14 +132,72 @@ void Thermometer::generate_thermometer(const type::Route* route) {
...
+
+            // We don't add a vertex twice
+            auto it = vertex_map.find(sp);
+            if (it == vertex_map.end()) {
+                current_vx = boost::add_vertex(VertexProperties(sp), stop_point_graph);
+                vertex_map[sp] = current_vx;
+            } else {
+	        current_vx = it->second;
+	    }
+	    if (!first) {
```
Comment (xlqian) [inline source/time_tables/thermometer.cpp]: To avoid creating a branch you may call std::slice

Comment (xlqian) [inline source/time_tables/thermometer.cpp]: sorry... it's for valarray

Comment (TeXitoi) [inline source/time_tables/thermometer.cpp]: or not ;-) std::slice is for valarray

Comment (xlqian) [inline source/time_tables/thermometer.cpp]: We can still do something pretty to avoid the branch :)

Comment (xavierraffin) [inline source/time_tables/thermometer.cpp]: Sorry I don't get it.

Do you want I change something or not ?

File: source/time_tables/thermometer.cpp [hunk 11]
```diff
@@ -130,14 +132,72 @@ void Thermometer::generate_thermometer(const type::Route* route) {
...
+	        boost::add_edge(previous_vx, current_vx, stop_point_graph);
+	    }
+	    first = false;
+	    previous_vx = current_vx;
+        }
+    }
+
+    std::vector< Vertex > c;
```
Comment (xlqian) [inline source/time_tables/thermometer.cpp]: please remove spaces around `Vertex`

File: source/time_tables/thermometer.cpp (hunk 10 above)
Comment (xavierraffin) [inline source/time_tables/thermometer.cpp]: Do you prefer a map of pointers ?

Else, there is a solution with labeled_graph
But that sound tricky and slow in comparaison to my O(1) map index :)

File: source/time_tables/thermometer.cpp (hunk 6 above)
Comment (xlqian) [inline source/time_tables/thermometer.cpp]: in fact, where `req` is empty or not, you push_back(v), right?

Comment (xlqian) [inline source/time_tables/thermometer.cpp]: you can do `continue` after the first push_back and remove `else`

File: source/time_tables/thermometer.cpp [hunk 12]
```diff
@@ -191,14 +251,14 @@ std::vector<uint32_t> Thermometer::stop_times_order_helper(const vector_idx &sto
...
+    if (journey_patterns[0].empty() && journey_patterns[1].empty())
         return {};
 
 
     //Si la journey_pattern une est vide, ou bien si le dernier de la journey_pattern n'est pas présent dans la journey_pattern 0, on renvoie la tete de la journey_pattern 1
-    if(journey_patterns[0].empty() || (!journey_patterns[1].empty() && pre_computed_lb[0][journey_patterns[1].back()] == 0) ) {
+    if (journey_patterns[0].empty() || (!journey_patterns[1].empty() && pre_computed_lb[0][journey_patterns[1].back()] == 0) ) {
         return {journey_patterns[1].back()};
-    } else if(journey_patterns[1].empty() ||(!journey_patterns[0].empty() && pre_computed_lb[1][journey_patterns[0].back()] == 0)) { //Même chose mais avec la journey_pattern 0
+    } else if (journey_patterns[1].empty() ||(!journey_patterns[0].empty() && pre_computed_lb[1][journey_patterns[0].back()] == 0)) { //Même chose mais avec la journey_pattern 0
```
Comment (xlqian) [inline source/time_tables/thermometer.cpp]: space after `||` please

**Comment by TeXitoi** [comment]
OK for me

**Comment by xlqian** [comment]
ok for me

**Comment by antoine-de** [comment]
Jenkins, test this please

**Comment by antoine-de** [comment]
thanks @xavierraffin for this !

**Comment by TeXitoi** [comment]
fix #1159

**Pull request merged**